multidim-interop/rust/src/bin/testplan_0500.rs

```diff
@@ -1,144 +1,148 @@
-        _ => panic!("Unsupported secure channel"),
-    };
-
-    trait AsyncRW: 'static + AsyncRead + AsyncWrite + Unpin + Send {}
-    impl<T> AsyncRW for T where T: 'static + AsyncRead + AsyncWrite + Unpin + Send {}
-
-    let f = |x: EitherOutput<
-        (libp2p::PeerId, NoiseOutput<core::Negotiated<C>>),
-        (libp2p::PeerId, TlsStream<core::Negotiated<C>>),
-    >|
-     -> (PeerId, Box<dyn AsyncRW>) {
-        match x {
-            EitherOutput::First((p_id, out)) => (p_id, Box::new(out)),
-            EitherOutput::Second((p_id, out)) => (p_id, Box::new(out)),
-        }
-    };
-
-    let secure_upgrade = secure_upgrade.map_outbound(f).map_inbound(f);
-    let authenticated = builder.authenticate(secure_upgrade);
-
-    let mux_upgrade = match muxer_param.as_str() {
-        "yamux" => EitherUpgrade::A(yamux::YamuxConfig::default()),
-        "mplex" => EitherUpgrade::B(mplex::MplexConfig::default()),
-        _ => panic!("Unsupported muxer"),
+    let mux_upgrade = match muxer_param {
+        Muxer::Yamux => EitherUpgrade::A(yamux::YamuxConfig::default()),
+        Muxer::Mplex => EitherUpgrade::B(mplex::MplexConfig::default()),
+        muxer => panic!("muxer {muxer:?} not supported for build_builder"),
     };
 
-    authenticated
-        .multiplex(mux_upgrade)
-        .timeout(Duration::from_secs(5))
-        .boxed()
+    let timeout = Duration::from_secs(5);
+
+    match secure_channel_param {
+        SecProtocol::Noise => builder
+            .authenticate(noise::NoiseAuthenticated::xx(&local_key).unwrap())
+            .multiplex(mux_upgrade)
+            .timeout(timeout)
+            .boxed(),
+        SecProtocol::Tls => builder
+            .authenticate(libp2p::tls::Config::new(&local_key).unwrap())
+            .multiplex(mux_upgrade)
+            .timeout(timeout)
+            .boxed(),
+    }
 }
 
-#[async_std::main]
+#[tokio::main]
 async fn main() -> Result<()> {
     let local_key = identity::Keypair::generate_ed25519();
     let local_peer_id = PeerId::from(local_key.public());
 
-    let transport_param =
-        env::var("transport").context("transport environment variable is not set")?;
-    let secure_channel_param =
-        env::var("security").context("security environment variable is not set")?;
-    let muxer_param = env::var("muxer").context("muxer environment variable is not set")?;
+    let transport_param: testplan::Transport = env::var("transport")
+        .context("transport environment variable is not set")?
+        .parse()
+        .context("unsupported transport")?;
+
+    let secure_channel_param: testplan::SecProtocol = env::var("security")
+        .context("security environment variable is not set")?
+        .parse()
+        .context("unsupported secure channel")?;
+
+    let muxer_param: Muxer = env::var("muxer")
+        .context("muxer environment variable is not set")?
+        .parse()
+        .context("unsupported muxer")?;
```

Comment:
Until we move the test to our repository (where we deduplicate the test using Git revisions), I don't think it makes much sense to be typed here because we have to repeat it in every test.

Reply:
it's now all on the testplans lib, which we can depend on the `rust-libp2p` repo and use from there wdyt?